### PR TITLE
Add alarm function backend

### LINKED ISSUE #19: alarm for the reminder
As a user,
I want to set alarms for my reminders,
so that i can be notified at specific times about import tasks or events.

- Research & Learning
-Backend Code
-Unit Tests
-Integration Tests
-Frontend Code
-------------------------------------------------------------
-Revise the bug of "Logout url" from "/logout" to "/api/logout"

### PR DESCRIPTION
**Pull request by TingSHI2015**
-closes #19

### CODE REVIEW

File: backend/src/main/java/com/github/tingshi2015/backend/reminder/ReminderService.java [hunk 1]
```diff
@@ -46,8 +49,19 @@ public Reminder updateAReminder(ReminderDTO updateReminder, String id) {
...
+
+    public List<Reminder> getUpcomingReminders() {
+        LocalDate today = LocalDate.now();
+        LocalTime now = LocalTime.now().withSecond(0).withNano(0); //ignore second & Nano-second
+
+        return reminderRepository.findAll().stream()
+                .filter(reminder -> reminder.date() != null && reminder.time() != null)
+                .filter(reminder -> reminder.date().equals(today) && reminder.time().withSecond(0).withNano(0).equals(now))
+                .collect(Collectors.toList());
+    }
```
Comment (josch87) [inline backend/src/main/java/com/github/tingshi2015/backend/reminder/ReminderService.java:66]: This is very inefficient. You request all reminders from the database and store them in the heap and then filter over them just to get some of them. This slows down your application and may crash when there are many reminders at some point by running out of memory. 
You should query your database directly with a method which only retrieves the necessary reminders, for example:

```
return reminderRepository.findRemindersByDateAndTime(today, now);
```

and in your repository your query:
```
@Query("{ 'date' : ?0, 'time' : ?1 }")
    List<Reminder> findRemindersByDateAndTime(LocalDate date, LocalTime time);
```

File: frontend/src/components/UpcomingReminderGallery.tsx [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+        axios.get("/api/reminders/upcoming")
+            .then(response => {
+                if(response.data.length > 0) {
+                    //setUpcomingReminders(response.data);
+                    setUpcomingReminders(preReminders => [...preReminders, ...response.data]);
+                    setShowUpcomingReminder(true);
+                }
+            })
+            .catch(error => console.error("Error getting upcoming reminders", error))
+            .finally(() => {console.log("getUpcomingReminder_successful")})
```
Comment (josch87) [inline frontend/src/components/UpcomingReminderGallery.tsx:32]: You should remove this line. If you want to you can implement frontend logging with info or debug level.

File: frontend/src/page/HomePage.tsx [hunk 3]
```diff
@@ -23,6 +24,8 @@ export default function HomePage(props: Readonly<HomePageProps>){
         setQuery(event.target.value)
     }
 
+    console.log(props)
```
Comment (josch87) [inline frontend/src/page/HomePage.tsx:27]:
```suggestion
```

**Review by josch87 — CHANGES_REQUESTED** [review]

File: backend/src/main/java/com/github/tingshi2015/backend/reminder/ReminderService.java (hunk 1 above)
Comment (TingSHI2015) [reply backend/src/main/java/com/github/tingshi2015/backend/reminder/ReminderService.java:66]: ganz gut tip! danke sehr :)

File: backend/src/main/java/com/github/tingshi2015/backend/reminder/NotificationService.java [hunk 4]
```diff
@@ -0,0 +1,10 @@
+package com.github.tingshi2015.backend.reminder;
+
+import org.springframework.stereotype.Service;
+
+@Service
+public class NotificationService {
+    public void sendNotification(Reminder reminder){
+        System.out.println("Reminder Alert: " + reminder.name() + " at " + reminder.time() + ", " + reminder.date());
```
[the remark this answers is not in the document]
Comment (TingSHI2015) [reply backend/src/main/java/com/github/tingshi2015/backend/reminder/NotificationService.java:8]: gut tip:)

**Pull request merged**